# 1.2 Cleaning and Scaling Data Set

## Contents
### 1. Import libraries and data set
### 2. Understand data set
### 3. Clean data set
### 4. Scaling
### 5. Export scaled data set

## 1. Import libraries and data set

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import os
import sklearn
from sklearn.preprocessing import StandardScaler

In [2]:
# create path to project folder
path = r'/Users/susanwang/Documents/CF_ML/ML_Project'

In [3]:
# open data set
df = pd.read_csv(os.path.join(path, 'Data Sets', 'Dataset-weather-prediction-dataset-processed.csv'), index_col=False)

### 2. Understand data set

In [4]:
df.head()

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0  19600101      1                  7               2.1            0.85   
1  19600102      1                  6               2.1            0.84   
2  19600103      1                  8               2.1            0.90   
3  19600104      1                  3               2.1            0.92   
4  19600105      1                  6               2.1            0.95   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0           1.018                    0.32                 0.09   
1           1.018                    0.36                 1.05   
2           1.018                    0.18                 0.30   
3           1.018                    0.58                 0.00   
4           1.018                    0.65                 0.14   

   BASEL_snow_depth  BASEL_sunshine  ...  VALENTIA_cloud_cover  \
0                 0             0.7  ...                     5   
1                 0             1.1  ...                     7   
2                 0             0.0  ...                     7   
3                 0             4.1  ...                     7   
4                 0             5.4  ...                     3   

   VALENTIA_humidity  VALENTIA_pressure  VALENTIA_global_radiation  \
0               0.88             1.0003                       0.45   
1               0.91             1.0007                       0.25   
2               0.91             1.0096                       0.17   
3               0.86             1.0184                       0.13   
4               0.80             1.0328                       0.46   

   VALENTIA_precipitation  VALENTIA_snow_depth  VALENTIA_sunshine  \
0                    0.34                    0                4.7   
1                    0.84                    0                0.7   
2                    0.08                    0                0.1   
3                    0.98                    0                0.0   
4                    0.00                    0                5.7   

   VALENTIA_temp_mean  VALENTIA_temp_min  VALENTIA_temp_max  
0                 8.5                6.0               10.9  
1                 8.9                5.6               12.1  
2                10.5                8.1               12.9  
3                 7.4                7.3               10.6  
4                 5.7                3.0                8.4  

[5 rows x 170 columns]

In [5]:
df.tail()

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
22945  20221027     10                  1               2.1            0.79   
22946  20221028     10                  6               2.1            0.77   
22947  20221029     10                  4               2.1            0.76   
22948  20221030     10                  5               2.1            0.80   
22949  20221031     10                  5               2.1            0.84   

       BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
22945          1.0248                    1.34                 0.22   
22946          1.0244                    1.34                 0.22   
22947          1.0227                    1.34                 0.22   
22948          1.0212                    1.34                 0.22   
22949          1.0193                    1.34                 0.22   

       BASEL_snow_depth  BASEL_sunshine  ...  VALENTIA_cloud_cover  \
22945                 0             7.7  ...                     5   
22946                 0             5.4  ...                     5   
22947                 0             6.1  ...                     5   
22948                 0             5.8  ...                     5   
22949                 0             3.2  ...                     5   

       VALENTIA_humidity  VALENTIA_pressure  VALENTIA_global_radiation  \
22945               0.82             1.0142                       1.13   
22946               0.82             1.0142                       1.13   
22947               0.82             1.0142                       1.13   
22948               0.82             1.0142                       1.13   
22949               0.82             1.0142                       1.13   

       VALENTIA_precipitation  VALENTIA_snow_depth  VALENTIA_sunshine  \
22945                    0.41                    0                3.4   
22946                    0.41                    0                3.4   
22947                    0.41                    0                3.4   
22948                    0.41                    0                3.4   
22949                    0.41                    0                3.4   

       VALENTIA_temp_mean  VALENTIA_temp_min  VALENTIA_temp_max  
22945                10.7                7.9               13.5  
22946                10.7                7.9               13.5  
22947                10.7                7.9               13.5  
22948                10.7                7.9               13.5  
22949                10.7                7.9               13.5  

[5 rows x 170 columns]

In [6]:
df.columns

Index(['DATE', 'MONTH', 'BASEL_cloud_cover', 'BASEL_wind_speed',
       'BASEL_humidity', 'BASEL_pressure', 'BASEL_global_radiation',
       'BASEL_precipitation', 'BASEL_snow_depth', 'BASEL_sunshine',
       ...
       'VALENTIA_cloud_cover', 'VALENTIA_humidity', 'VALENTIA_pressure',
       'VALENTIA_global_radiation', 'VALENTIA_precipitation',
       'VALENTIA_snow_depth', 'VALENTIA_sunshine', 'VALENTIA_temp_mean',
       'VALENTIA_temp_min', 'VALENTIA_temp_max'],
      dtype='object', length=170)

170 columns, presumably the weather variables for the 18 different weather stations in Europe. Date ranges from 1960 to 2022.

In [7]:
df.shape

(22950, 170)

## 3. Clean data set

### Check for missing values

In [8]:
df.isnull().sum()

DATE                   0
MONTH                  0
BASEL_cloud_cover      0
BASEL_wind_speed       0
BASEL_humidity         0
                      ..
VALENTIA_snow_depth    0
VALENTIA_sunshine      0
VALENTIA_temp_mean     0
VALENTIA_temp_min      0
VALENTIA_temp_max      0
Length: 170, dtype: int64

In [9]:
# try to display all answers
pd.options.display.max_rows = 200
df.isnull().sum()

DATE                           0
MONTH                          0
BASEL_cloud_cover              0
BASEL_wind_speed               0
BASEL_humidity                 0
BASEL_pressure                 0
BASEL_global_radiation         0
BASEL_precipitation            0
BASEL_snow_depth               0
BASEL_sunshine                 0
BASEL_temp_mean                0
BASEL_temp_min                 0
BASEL_temp_max                 0
BELGRADE_cloud_cover           0
BELGRADE_humidity              0
BELGRADE_pressure              0
BELGRADE_global_radiation      0
BELGRADE_precipitation         0
BELGRADE_sunshine              0
BELGRADE_temp_mean             0
BELGRADE_temp_min              0
BELGRADE_temp_max              0
BUDAPEST_cloud_cover           0
BUDAPEST_humidity              0
BUDAPEST_pressure              0
BUDAPEST_global_radiation      0
BUDAPEST_precipitation         0
BUDAPEST_sunshine              0
BUDAPEST_temp_mean             0
BUDAPEST_temp_min              0
BUDAPEST_t

No missing values.

### Check for duplicates

In [10]:
df_dups = df[df.duplicated()]
df_dups

Empty DataFrame
Columns: [DATE, MONTH, BASEL_cloud_cover, BASEL_wind_speed, BASEL_humidity, BASEL_pressure, BASEL_global_radiation, BASEL_precipitation, BASEL_snow_depth, BASEL_sunshine, BASEL_temp_mean, BASEL_temp_min, BASEL_temp_max, BELGRADE_cloud_cover, BELGRADE_humidity, BELGRADE_pressure, BELGRADE_global_radiation, BELGRADE_precipitation, BELGRADE_sunshine, BELGRADE_temp_mean, BELGRADE_temp_min, BELGRADE_temp_max, BUDAPEST_cloud_cover, BUDAPEST_humidity, BUDAPEST_pressure, BUDAPEST_global_radiation, BUDAPEST_precipitation, BUDAPEST_sunshine, BUDAPEST_temp_mean, BUDAPEST_temp_min, BUDAPEST_temp_max, DEBILT_cloud_cover, DEBILT_wind_speed, DEBILT_humidity, DEBILT_pressure, DEBILT_global_radiation, DEBILT_precipitation, DEBILT_sunshine, DEBILT_temp_mean, DEBILT_temp_min, DEBILT_temp_max, DUSSELDORF_cloud_cover, DUSSELDORF_wind_speed, DUSSELDORF_humidity, DUSSELDORF_pressure, DUSSELDORF_global_radiation, DUSSELDORF_precipitation, DUSSELDORF_snow_depth, DUSSELDORF_sunshine, DUSSELDORF_temp_mean, DUSSELDORF_temp_min, DUSSELDORF_temp_max, GDANSK_cloud_cover, GDANSK_humidity, GDANSK_precipitation, GDANSK_snow_depth, GDANSK_temp_mean, GDANSK_temp_min, GDANSK_temp_max, HEATHROW_cloud_cover, HEATHROW_humidity, HEATHROW_pressure, HEATHROW_global_radiation, HEATHROW_precipitation, HEATHROW_snow_depth, HEATHROW_sunshine, HEATHROW_temp_mean, HEATHROW_temp_min, HEATHROW_temp_max, KASSEL_wind_speed, KASSEL_humidity, KASSEL_pressure, KASSEL_global_radiation, KASSEL_precipitation, KASSEL_sunshine, KASSEL_temp_mean, KASSEL_temp_min, KASSEL_temp_max, LJUBLJANA_cloud_cover, LJUBLJANA_wind_speed, LJUBLJANA_humidity, LJUBLJANA_pressure, LJUBLJANA_global_radiation, LJUBLJANA_precipitation, LJUBLJANA_sunshine, LJUBLJANA_temp_mean, LJUBLJANA_temp_min, LJUBLJANA_temp_max, MAASTRICHT_cloud_cover, MAASTRICHT_wind_speed, MAASTRICHT_humidity, MAASTRICHT_pressure, MAASTRICHT_global_radiation, MAASTRICHT_precipitation, MAASTRICHT_sunshine, MAASTRICHT_temp_mean, MAASTRICHT_temp_min, MAASTRICHT_temp_max, MADRID_cloud_cover, MADRID_wind_speed, ...]
Index: []

[0 rows x 170 columns]

No duplicates.

## Checking for outliers

In [12]:
pd.options.display.max_columns = None

In [13]:
df.describe()

DATE         MONTH  BASEL_cloud_cover  BASEL_wind_speed  \
count  2.295000e+04  22950.000000       22950.000000      22950.000000   
mean   1.990984e+07      6.509630           5.410763          2.120462   
std    1.813833e+05      3.443672           2.406115          0.732625   
min    1.960010e+07      1.000000           0.000000          0.000000   
25%    1.975092e+07      4.000000           4.000000          2.100000   
50%    1.991060e+07      7.000000           6.000000          2.100000   
75%    2.007021e+07      9.000000           7.000000          2.100000   
max    2.022103e+07     12.000000           8.000000         16.300000   

       BASEL_humidity  BASEL_pressure  BASEL_global_radiation  \
count    22950.000000    22950.000000            22950.000000   
mean         0.758554        1.018013                1.345244   
std          0.110699        0.006543                0.931158   
min          0.350000        0.974700                0.010000   
25%          0.680000        1.015800                0.540000   
50%          0.770000        1.018000                1.130000   
75%          0.840000        1.020100                2.070000   
max          1.000000        1.045200                4.560000   

       BASEL_precipitation  BASEL_snow_depth  BASEL_sunshine  BASEL_temp_mean  \
count         22950.000000      22950.000000    22950.000000     22950.000000   
mean              0.222305          0.359564        4.592222        10.392471   
std               0.498995          2.006231        4.310808         7.363575   
min               0.000000         -2.000000        0.000000       -18.500000   
25%               0.000000          0.000000        0.400000         4.700000   
50%               0.000000          0.000000        3.600000        10.700000   
75%               0.220000          0.000000        7.900000        16.200000   
max               8.500000         49.000000       16.800000        29.200000   

       BASEL_temp_min  BASEL_temp_max  BELGRADE_cloud_cover  \
count    22950.000000    22950.000000          22950.000000   
mean         6.317015       15.076322              4.244357   
std          6.524121        8.730759              2.689274   
min        -23.300000      -14.000000              0.000000   
25%          1.400000        8.300000              2.000000   
50%          6.500000       15.400000              4.000000   
75%         11.500000       21.900000              7.000000   
max         22.400000       38.600000             10.000000   

       BELGRADE_humidity  BELGRADE_pressure  BELGRADE_global_radiation  \
count       22950.000000       22950.000000               22950.000000   
mean            0.683002           1.017073                   1.573360   
std             0.140290           0.007543                   0.933023   
min             0.230000           0.978700                   0.120000   
25%             0.580000           1.012400                   0.750000   
50%             0.680000           1.016600                   1.570000   
75%             0.790000           1.021400                   2.360000   
max             0.990000           1.046000                   3.810000   

       BELGRADE_precipitation  BELGRADE_sunshine  BELGRADE_temp_mean  \
count            22950.000000       22950.000000        22950.000000   
mean                 0.189349           5.770780           12.659935   
std                  0.509952           4.328599            8.811427   
min                  0.000000           0.000000          -17.300000   
25%                  0.000000           1.300000            5.700000   
50%                  0.000000           5.700000           13.400000   
75%                  0.100000           9.500000           19.800000   
max                 10.980000          14.300000           34.600000   

       BELGRADE_temp_min  BELGRADE_temp_max  BUDAPEST_cloud_cover  \
count       22950.000000       22950.000000          22950.000000   
mean      

In [22]:
# using IQR method
for col in df.columns:
    Q1 = np.percentile(df[col], 25, method='midpoint')
    Q3 = np.percentile(df[col], 75, method='midpoint')
    IQR = Q3 - Q1
    
    upper = Q3 + 1.5*IQR
    lower = Q1 - 1.5*IQR
    upper_array = np.array(df[col] >= upper)
    lower_array = np.array(df[col] <= lower)
    # print if the number of outliers exceed 5% of the total number of observations:
    if upper_array.sum() > 1147 or lower_array.sum() > 1147:
        print(f'lower {lower} upper {upper}') 
        print(f'{col} has {upper_array.sum()} upper outliers and {lower_array.sum()} lower outliers.')

lower 2.1 upper 2.1
BASEL_wind_speed has 18035 upper outliers and 19917 lower outliers.
lower 1.00935 upper 1.0265499999999999
BASEL_pressure has 2119 upper outliers and 1875 lower outliers.
lower -0.33 upper 0.55
BASEL_precipitation has 3046 upper outliers and 0 lower outliers.
lower 0.0 upper 0.0
BASEL_snow_depth has 22946 upper outliers and 21429 lower outliers.
lower -0.15000000000000002 upper 0.25
BELGRADE_precipitation has 4156 upper outliers and 0 lower outliers.
lower 1.0 upper 9.0
BUDAPEST_cloud_cover has 0 upper outliers and 1896 lower outliers.
lower 0.51 upper 0.8299999999999998
BUDAPEST_humidity has 2244 upper outliers and 2037 lower outliers.
lower 1.0100499999999997 upper 1.02325
BUDAPEST_pressure has 2841 upper outliers and 2429 lower outliers.
lower -0.12 upper 0.2
BUDAPEST_precipitation has 3704 upper outliers and 0 lower outliers.
lower -0.375 upper 0.625
DEBILT_precipitation has 2816 upper outliers and 0 lower outliers.
lower -0.36 upper 0.6
DUSSELDORF_precipitation

There is a disparate amount of outliers for each variable, some with a very large number of outliers. 
There are too many to consider handling, but this would be a good reference when trying machine learning later.
Snow depth and precipitation seem to have consistently high numbers of outliers. 

## 4. Scaling

In [11]:
df.dtypes

DATE                             int64
MONTH                            int64
BASEL_cloud_cover                int64
BASEL_wind_speed               float64
BASEL_humidity                 float64
BASEL_pressure                 float64
BASEL_global_radiation         float64
BASEL_precipitation            float64
BASEL_snow_depth                 int64
BASEL_sunshine                 float64
BASEL_temp_mean                float64
BASEL_temp_min                 float64
BASEL_temp_max                 float64
BELGRADE_cloud_cover             int64
BELGRADE_humidity              float64
BELGRADE_pressure              float64
BELGRADE_global_radiation      float64
BELGRADE_precipitation         float64
BELGRADE_sunshine              float64
BELGRADE_temp_mean             float64
BELGRADE_temp_min              float64
BELGRADE_temp_max              float64
BUDAPEST_cloud_cover             int64
BUDAPEST_humidity              float64
BUDAPEST_pressure              float64
BUDAPEST_global_radiation

All columns are numeric, but the date and month may not be useful anymore when they are scaled. 
But, looking ahead at the next tasks, I realize we do need the dates to filter the year for analysis. 

### Scale

In [12]:
# create a list of num columns
columns = df.columns.tolist()
num_col = columns[2:]
print(len(num_col))
num_col

168


['BASEL_cloud_cover',
 'BASEL_wind_speed',
 'BASEL_humidity',
 'BASEL_pressure',
 'BASEL_global_radiation',
 'BASEL_precipitation',
 'BASEL_snow_depth',
 'BASEL_sunshine',
 'BASEL_temp_mean',
 'BASEL_temp_min',
 'BASEL_temp_max',
 'BELGRADE_cloud_cover',
 'BELGRADE_humidity',
 'BELGRADE_pressure',
 'BELGRADE_global_radiation',
 'BELGRADE_precipitation',
 'BELGRADE_sunshine',
 'BELGRADE_temp_mean',
 'BELGRADE_temp_min',
 'BELGRADE_temp_max',
 'BUDAPEST_cloud_cover',
 'BUDAPEST_humidity',
 'BUDAPEST_pressure',
 'BUDAPEST_global_radiation',
 'BUDAPEST_precipitation',
 'BUDAPEST_sunshine',
 'BUDAPEST_temp_mean',
 'BUDAPEST_temp_min',
 'BUDAPEST_temp_max',
 'DEBILT_cloud_cover',
 'DEBILT_wind_speed',
 'DEBILT_humidity',
 'DEBILT_pressure',
 'DEBILT_global_radiation',
 'DEBILT_precipitation',
 'DEBILT_sunshine',
 'DEBILT_temp_mean',
 'DEBILT_temp_min',
 'DEBILT_temp_max',
 'DUSSELDORF_cloud_cover',
 'DUSSELDORF_wind_speed',
 'DUSSELDORF_humidity',
 'DUSSELDORF_pressure',
 'DUSSELDORF_global_

In [13]:
# create scaler object
scaler = StandardScaler()

In [14]:
# create a copy of original df
df_scaled = df.copy()
df_scaled.head()

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0  19600101      1                  7               2.1            0.85   
1  19600102      1                  6               2.1            0.84   
2  19600103      1                  8               2.1            0.90   
3  19600104      1                  3               2.1            0.92   
4  19600105      1                  6               2.1            0.95   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0           1.018                    0.32                 0.09   
1           1.018                    0.36                 1.05   
2           1.018                    0.18                 0.30   
3           1.018                    0.58                 0.00   
4           1.018                    0.65                 0.14   

   BASEL_snow_depth  BASEL_sunshine  ...  VALENTIA_cloud_cover  \
0                 0             0.7  ...                     5   
1                 0             1.1  ...                     7   
2                 0             0.0  ...                     7   
3                 0             4.1  ...                     7   
4                 0             5.4  ...                     3   

   VALENTIA_humidity  VALENTIA_pressure  VALENTIA_global_radiation  \
0               0.88             1.0003                       0.45   
1               0.91             1.0007                       0.25   
2               0.91             1.0096                       0.17   
3               0.86             1.0184                       0.13   
4               0.80             1.0328                       0.46   

   VALENTIA_precipitation  VALENTIA_snow_depth  VALENTIA_sunshine  \
0                    0.34                    0                4.7   
1                    0.84                    0                0.7   
2                    0.08                    0                0.1   
3                    0.98                    0                0.0   
4                    0.00                    0                5.7   

   VALENTIA_temp_mean  VALENTIA_temp_min  VALENTIA_temp_max  
0                 8.5                6.0               10.9  
1                 8.9                5.6               12.1  
2                10.5                8.1               12.9  
3                 7.4                7.3               10.6  
4                 5.7                3.0                8.4  

[5 rows x 170 columns]

In [15]:
# scale num columns on copied df
df_scaled[num_col] = scaler.fit_transform(df_scaled[num_col])

In [16]:
df_scaled.head()

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0  19600101      1           0.660514          -0.02793        0.826097   
1  19600102      1           0.244897          -0.02793        0.735760   
2  19600103      1           1.076130          -0.02793        1.277781   
3  19600104      1          -1.001953          -0.02793        1.458455   
4  19600105      1           0.244897          -0.02793        1.729466   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0       -0.001949               -1.101066            -0.265148   
1       -0.001949               -1.058108             1.658760   
2       -0.001949               -1.251420             0.155707   
3       -0.001949               -0.821838            -0.445514   
4       -0.001949               -0.746661            -0.164944   

   BASEL_snow_depth  BASEL_sunshine  ...  VALENTIA_cloud_cover  \
0         -0.179228       -0.902918  ...             -0.443701   
1         -0.179228       -0.810126  ...              0.783085   
2         -0.179228       -1.065304  ...              0.783085   
3         -0.179228       -0.114186  ...              0.783085   
4         -0.179228        0.187388  ...             -1.670486   

   VALENTIA_humidity  VALENTIA_pressure  VALENTIA_global_radiation  \
0           0.761754          -1.299744                  -0.806427   
1           1.183580          -1.262455                  -1.042055   
2           1.183580          -0.432779                  -1.136306   
3           0.480538           0.387574                  -1.183432   
4          -0.363113           1.729970                  -0.794645   

   VALENTIA_precipitation  VALENTIA_snow_depth  VALENTIA_sunshine  \
0               -0.088407            -0.024706           0.372147   
1                0.503361            -0.024706          -0.829285   
2               -0.396127            -0.024706          -1.009500   
3                0.669056            -0.024706          -1.039536   
4               -0.490810            -0.024706           0.672505   

   VALENTIA_temp_mean  VALENTIA_temp_min  VALENTIA_temp_max  
0           -0.668215          -0.519743          -0.752237  
1           -0.548046          -0.629054          -0.407141  
2           -0.067372           0.054135          -0.177078  
3           -0.998679          -0.164486          -0.838511  
4           -1.509396          -1.339569          -1.471186  

[5 rows x 170 columns]

In [17]:
# compare to original
df.head()

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0  19600101      1                  7               2.1            0.85   
1  19600102      1                  6               2.1            0.84   
2  19600103      1                  8               2.1            0.90   
3  19600104      1                  3               2.1            0.92   
4  19600105      1                  6               2.1            0.95   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0           1.018                    0.32                 0.09   
1           1.018                    0.36                 1.05   
2           1.018                    0.18                 0.30   
3           1.018                    0.58                 0.00   
4           1.018                    0.65                 0.14   

   BASEL_snow_depth  BASEL_sunshine  ...  VALENTIA_cloud_cover  \
0                 0             0.7  ...                     5   
1                 0             1.1  ...                     7   
2                 0             0.0  ...                     7   
3                 0             4.1  ...                     7   
4                 0             5.4  ...                     3   

   VALENTIA_humidity  VALENTIA_pressure  VALENTIA_global_radiation  \
0               0.88             1.0003                       0.45   
1               0.91             1.0007                       0.25   
2               0.91             1.0096                       0.17   
3               0.86             1.0184                       0.13   
4               0.80             1.0328                       0.46   

   VALENTIA_precipitation  VALENTIA_snow_depth  VALENTIA_sunshine  \
0                    0.34                    0                4.7   
1                    0.84                    0                0.7   
2                    0.08                    0                0.1   
3                    0.98                    0                0.0   
4                    0.00                    0                5.7   

   VALENTIA_temp_mean  VALENTIA_temp_min  VALENTIA_temp_max  
0                 8.5                6.0               10.9  
1                 8.9                5.6               12.1  
2                10.5                8.1               12.9  
3                 7.4                7.3               10.6  
4                 5.7                3.0                8.4  

[5 rows x 170 columns]

## 5. Export scaled data set

In [18]:
df_scaled.to_csv(os.path.join(path, 'Data Sets', 'weather_dataset_scaled.csv'), index=False)